### PR TITLE
r/elasticsearch_domain - tag on create + refactor to use waiter and finder packages

### LINKED ISSUE #20544: resource/aws_elasticsearch_domain: Now supports tagging on creation
### Description

`aws_elasticsearch_domain` currently adds tags in a separate API call during creation. The `elasticsearch.CreateElasticsearchDomainInput` struct has a `TagList` field to allow a single call to the API.

### Affected Resource

* aws_elasticsearch_domain

### References

* #20498 is about a tagging issue with `aws_elasticsearch_domain`

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #20544

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSElasticSearchDomain_'
--- PASS: TestAccElasticsearchDomain_WithVolumeType_missing (877.51s)
--- PASS: TestAccElasticsearchDomain_nodeToNodeEncryption (900.18s)
--- PASS: TestAccElasticsearchDomain_policy (926.87s)
--- PASS: TestAccElasticsearchDomain_LogPublishingOptions_searchSlowLogs (927.84s)
--- PASS: TestAccElasticsearchDomain_duplicate (529.13s)
--- PASS: TestAccElasticsearchDomain_LogPublishingOptions_esApplicationLogs (989.10s)
--- PASS: TestAccElasticsearchDomain_LogPublishingOptions_auditLogs (1030.62s)
--- PASS: TestAccElasticsearchDomain_update (1990.95s)
--- PASS: TestAccElasticsearchDomain_cognitoOptionsUpdate (1683.11s)
--- PASS: TestAccElasticsearchDomain_EncryptAtRestSpecify_key (1201.23s)
--- PASS: TestAccElasticsearchDomain_internetToVPCEndpoint (2659.77s)
--- PASS: TestAccElasticsearchDomain_cognitoOptionsCreateAndRemove (1794.31s)
--- PASS: TestAccElasticsearchDomain_UpdateVolume_type (2708.25s)
--- PASS: TestAccElasticsearchDomain_complex (914.07s)
--- PASS: TestAccElasticsearchDomain_Update_version (2998.24s)
--- PASS: TestAccElasticsearchDomain_v23 (659.78s)
--- PASS: TestAccElasticsearchDomain_AdvancedSecurityOptions_iam (1284.14s)
--- PASS: TestAccElasticsearchDomain_LogPublishingOptions_indexSlowLogs (1187.19s)
--- PASS: TestAccElasticsearchDomain_AdvancedSecurityOptions_userDB (1388.94s)
--- PASS: TestAccElasticsearchDomain_requireHTTPS (1343.83s)
--- PASS: TestAccElasticsearchDomain_customEndpoint (1352.31s)
--- PASS: TestAccElasticsearchDomain_Cluster_zoneAwareness (4028.55s)
```

### CODE REVIEW

**Comment by wajdiail** [comment]
Just checking when this MR will be merged. We are waiting for this feature

**Comment by mattheidelbaugh** [comment]
Seconded we are also waiting on this

**Comment by anGie44** [comment]
Hi @DrFaust92 👋 do you mind fixing up the merge conflicts that came up after the service packages refactor when you have a moment?

**Comment by DrFaust92** [comment]
@anGie44 rebased + more fixes but didnt get to run all tests yet.

**Comment by anGie44** [comment]
thank you @DrFaust92 ! No worries about running the tests, I can do so on my end 👍

**Comment by DrFaust92** [comment]
@anGie44 i see something else was merged and now it created a conflict. would like me to rebase and try to take care of it or you will look into it?

**Comment by anGie44** [comment]
Hey @DrFaust92 -- apologies for not getting back to this sooner and the merge conflicts! i'll have a look and rebase

File: internal/service/elasticsearch/domain_test.go [hunk 1]
```diff
@@ -1164,6 +1175,27 @@ func TestAccElasticsearchDomain_Update_version(t *testing.T) {
...
+		PreCheck:     func() { acctest.PreCheck(t); testAccPreCheckIamServiceLinkedRoleEs(t) },
+		ErrorCheck:   acctest.ErrorCheck(t, elasticsearch.EndpointsID),
+		Providers:    acctest.Providers,
+		CheckDestroy: testAccCheckDomainDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDomainConfig(rName),
+				Check: resource.ComposeTestCheckFunc(
+					acctest.CheckResourceDisappears(acctest.Provider, tfelasticsearch.ResourceDomain(), resourceName),
+					acctest.CheckResourceDisappears(acctest.Provider, tfelasticsearch.ResourceDomain(), resourceName)),
```
Comment (anGie44) [inline internal/service/elasticsearch/domain_test.go:1192]: Small change needed:  this like looks to be a duplicate

**Comment by anGie44** [comment]
Hi again @DrFaust92 -- so i rebased and merged back in the changes from https://github.com/hashicorp/terraform-provider-aws/pull/22157/files#diff-06ff9b94aba8bd561ec9bbcca77f52542eb0bc702f2b43c5c73f8a467797bd95 which added the new function `advancedOptionsIgnoreDefault` . originally, you updated the d.Set("advanced_options", ..._ to use the `flex` packages map flattener so if you think there is a bug there do you mind creating a new PR with that change and the test changes as it looks like `TestAccElasticsearchDomain_complex` is now failing with the following since the 2 advanced options are explicitly configured in the test config: 
```
=== RUN   TestAccElasticsearchDomain_complex
=== PAUSE TestAccElasticsearchDomain_complex
=== CONT  TestAccElasticsearchDomain_complex
    domain_test.go:379: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.
        
        (map[string]string) (len=1) {
         (string) (len=18) "advanced_options.%": (string) (len=1) "1"
        }
        
        
        (map[string]string) (len=3) {
         (string) (len=18) "advanced_options.%": (string) (len=1) "3",
         (string) (len=47) "advanced_options.override_main_response_version": (string) (len=5) "false",
         (string) (len=55) "advanced_options.rest.action.multi.allow_explicit_index": (string) (len=4) "true"
        }
--- FAIL: TestAccElasticsearchDomain_complex (1183.21s)
```

File: internal/service/elasticsearch/domain_test.go (hunk 1 above)
Comment (DrFaust92) [reply internal/service/elasticsearch/domain_test.go]: duplicate is on purpose to check the delete function is actually idempotent. (usually missing catching an error on delete) should probably be added into `CheckResourceDisappears` but could potentially break existing tests.

**Comment by DrFaust92** [comment]
@anGie44 i dont recall atm why i did it but i do remember failures around that test (and others) in regards to `advanced_options`. ill take a look again later but if the fixes you added make the test work as expected than its probably better than my changes 👍

**Comment by anGie44** [comment]
Thanks @DrFaust92 -- i think for the scope of this PR, I would just recommend removing the test config additions:
```
  "override_main_response_version"         = "false"
    "rest.action.multi.allow_explicit_index" = "true"
```
and then add a changelog entry and we should be good to go 👍

**Comment by anGie44** [comment]
Output of acceptance tests (commercial):
```
--- PASS: TestAccElasticsearchDomainDataSource_Data_advanced (1788.44s)
--- PASS: TestAccElasticsearchDomainDataSource_Data_basic (1046.39s)
--- PASS: TestAccElasticsearchDomainPolicy_basic (1356.43s)
--- PASS: TestAccElasticsearchDomainSamlOptions_Disabled (2650.71s)
--- PASS: TestAccElasticsearchDomainSamlOptions_Update (1878.07s)
--- PASS: TestAccElasticsearchDomainSamlOptions_basic (1983.97s)
--- PASS: TestAccElasticsearchDomainSamlOptions_disappears (1379.58s)
--- PASS: TestAccElasticsearchDomainSamlOptions_disappears_Domain (1434.56s)
--- PASS: TestAccElasticsearchDomain_AdvancedSecurityOptions_disabled (1172.60s)
--- PASS: TestAccElasticsearchDomain_AdvancedSecurityOptions_iam (1398.29s)
--- PASS: TestAccElasticsearchDomain_AdvancedSecurityOptions_userDB (1297.78s)
--- PASS: TestAccElasticsearchDomain_AutoTuneOptions (2080.80s)
--- PASS: TestAccElasticsearchDomain_Cluster_zoneAwareness (5685.77s)
--- PASS: TestAccElasticsearchDomain_EncryptAtRestDefault_key (1438.40s)
--- PASS: TestAccElasticsearchDomain_EncryptAtRestSpecify_key (2504.77s)
--- PASS: TestAccElasticsearchDomain_LogPublishingOptions_auditLogs (1447.30s)
--- PASS: TestAccElasticsearchDomain_LogPublishingOptions_esApplicationLogs (1714.02s)
--- PASS: TestAccElasticsearchDomain_LogPublishingOptions_indexSlowLogs (1027.60s)
--- PASS: TestAccElasticsearchDomain_LogPublishingOptions_searchSlowLogs (1670.37s)
--- PASS: TestAccElasticsearchDomain_UpdateVolume_type (3735.18s)
--- PASS: TestAccElasticsearchDomain_Update_version (3839.77s)
--- PASS: TestAccElasticsearchDomain_VPC_update (2738.44s)
--- PASS: TestAccElasticsearchDomain_WithVolumeType_missing (811.79s)
--- PASS: TestAccElasticsearchDomain_basic (1043.95s)
--- PASS: TestAccElasticsearchDomain_cognitoOptionsCreateAndRemove (2284.04s)
--- PASS: TestAccElasticsearchDomain_cognitoOptionsUpdate (2069.46s)
--- PASS: TestAccElasticsearchDomain_complex (1039.04s)
--- PASS: TestAccElasticsearchDomain_customEndpoint (2823.50s)
--- PASS: TestAccElasticsearchDomain_disappears (1473.62s)
--- PASS: TestAccElasticsearchDomain_duplicate (833.60s)
--- PASS: TestAccElasticsearchDomain_internetToVPCEndpoint (2390.55s)
--- PASS: TestAccElasticsearchDomain_nodeToNodeEncryption (943.10s)
--- PASS: TestAccElasticsearchDomain_policy (1183.45s)
--- PASS: TestAccElasticsearchDomain_policyIgnoreEquivalent (960.67s)
--- PASS: TestAccElasticsearchDomain_requireHTTPS (3572.18s)
--- PASS: TestAccElasticsearchDomain_tags (853.91s)
--- PASS: TestAccElasticsearchDomain_update (2370.28s)
--- PASS: TestAccElasticsearchDomain_v23 (1258.82s)
--- PASS: TestAccElasticsearchDomain_vpc (2983.09s)
--- PASS: TestAccElasticsearchDomain_warm (5328.24s)
--- PASS: TestAccElasticsearchDomain_withDedicatedMaster (4630.98s)
```

**Review by anGie44 — APPROVED** [review]
@DrFaust92 , i've reverted the test config back to the state of `main` and everything's passing again with the refactor changes. Thanks again for these changes!

**Pull request merged**